src/object/srv_ec_aggregate.c

```diff
@@ -983,34 +2093,76 @@ ds_obj_ec_aggregate(struct ds_cont_child *cont, daos_epoch_range_t *epr,
 	vos_iter_param_t	 iter_param = { 0 };
 	struct vos_iter_anchors  anchors = { 0 };
 	struct ec_agg_param	 agg_param = { 0 };
+	daos_handle_t		 ph = DAOS_HDL_INVAL;
+	int			*status;
 	int			 rc = 0;
 
-	/* The caller holds the pool reference (via @cont). */
-	agg_param.ap_pool_info.api_pool = cont->sc_pool->spc_pool;
+	uuid_copy(agg_param.ap_pool_info.api_pool_uuid,
+		  cont->sc_pool->spc_uuid);
 	uuid_copy(agg_param.ap_pool_info.api_cont_uuid, cont->sc_uuid);
-
-	agg_param.ap_pool_info.api_pool_version =
-		cont->sc_pool->spc_pool->sp_map_version;
-
+	agg_param.ap_pool_info.api_pool = cont->sc_pool->spc_pool;
 	agg_param.ap_cont_handle	= cont->sc_hdl;
-	agg_param.ap_credits_max	= EC_AGG_ITERATION_MAX;
 	agg_param.ap_yield_func		= yield_func;
 	agg_param.ap_yield_arg		= yield_arg;
+	agg_param.ap_credits_max	= EC_AGG_ITERATION_MAX;
+
+	rc = ABT_eventual_create(sizeof(*status),
+				 &agg_param.ap_pool_info.api_eventual);
+	if (rc != ABT_SUCCESS)
+		return dss_abterr2der(rc);
+	rc = dss_ult_create(agg_iv_ult, &agg_param, DSS_ULT_POOL_SRV, 0, 0,
+			    NULL);
+	if (rc)
+		goto out;
+	rc = ABT_eventual_wait(agg_param.ap_pool_info.api_eventual,
+			       (void **)&status);
+	if (rc != ABT_SUCCESS) {
+		rc = dss_abterr2der(rc);
+		goto out;
+	}
+	if (*status != 0) {
+		rc = *status;
+		goto out;
+	}
+
+	rc = dsc_pool_open(agg_param.ap_pool_info.api_pool_uuid,
+			   agg_param.ap_pool_info.api_poh_uuid, DAOS_PC_RW,
+			   NULL, agg_param.ap_pool_info.api_pool->sp_map,
+			   agg_param.ap_pool_info.api_svc_list, &ph);
+	if (rc) {
+		D_ERROR("dsc_pool_open failed: "DF_RC"\n", DP_RC(rc));
+		goto out;
+	}
+
+	rc = dsc_cont_open(ph, agg_param.ap_pool_info.api_cont_uuid,
+			   agg_param.ap_pool_info.api_coh_uuid, DAOS_COO_RW,
+			   &agg_param.ap_pool_info.api_cont_hdl);
+	if (rc) {
+		D_ERROR("dsc_cont_open failed: "DF_RC"\n", DP_RC(rc));
+		goto out;
+	}
+	D_INIT_LIST_HEAD(&agg_param.ap_agg_entry.ae_cur_stripe.as_dextents);
+	D_INIT_LIST_HEAD(&agg_param.ap_agg_entry.ae_cur_stripe.as_hoextents);
+
 	iter_param.ip_hdl		= cont->sc_hdl;
 	iter_param.ip_epr.epr_lo	= epr->epr_lo;
 	iter_param.ip_epr.epr_hi	= epr->epr_hi;
 	iter_param.ip_flags		= VOS_IT_RECX_VISIBLE;
-	iter_param.ip_epc_expr		= VOS_IT_EPC_RR;
 	iter_param.ip_recx.rx_idx	= 0ULL;
 	iter_param.ip_recx.rx_nr	= ~PARITY_INDICATOR;
-
-	D_INIT_LIST_HEAD(&agg_param.ap_agg_entry.ae_cur_stripe.as_dextents);
+	crt_group_rank(NULL, &myrank);
 
 	rc = vos_iterate(&iter_param, VOS_ITER_OBJ, true, &anchors,
 			 agg_iterate_pre_cb, agg_iterate_post_cb,
 			 &agg_param, NULL);
 
+	dsc_cont_close(ph, agg_param.ap_pool_info.api_cont_hdl);
+out:
+	daos_prop_entries_free(agg_param.ap_prop);
+	D_FREE(agg_param.ap_prop);
+	ABT_eventual_free(&agg_param.ap_pool_info.api_eventual);
 	agg_sgl_fini(&agg_param.ap_agg_entry.ae_sgl);
+	dsc_pool_close(ph);
```

Comment:
hmm, you need differentiate what needs to cleanup in case there is failure happens?

Reply:
The sgl_fini just calls D_FREE, which can be called on NULL pointers.  So I think this is okay as is, but I will check to make sure.

Reply:
It seems dsc_pool_close is not needed if eventual_wait or ult_create fails?

Hmm, it seems dsc_cont_close is also missing here?

Reply:
dsc_cont_close is on line 2159.